### ISSUE TITLE
compile matcher doesn't honor facts when selecting a default provider

### ISSUE THREAD

**Issue by geoffdavis** [rodjek/rspec-puppet#256]
Possibly related to GH-175, the compile matcher doesn't seem to take into account overridden facts when deciding which provider to use. I discovered this when testing a class that made use of the User type.

On my Mac, no matter what the osfamily fact was set to, the provider for Users was set to the OSX default of "directoryservice". Other matchers seem to work fine, such as "contain_user".

My workaround has been to set a pre_condition that overrides the automatic provider selection on a per-OS basis.

Here's a simple example that tests provider specific behavior. Running this on my Mac results in the Mac's directoryservice provider being used for the User type, despite the facts being set on the context.

``` puppet
class users::test {
  # Darwin cannot use managehome by default with directoryservice provider
  $my_managehome = $::osfamily ? {
    'Darwin' => false,
    default => true,
  }
  file { '/home': ensure => directory }
  user { 'testuser':
    managehome => $my_managehome,
    home => '/home/testuser',
  }
}
```

``` ruby
require 'puppetlabs_spec_helper/module_spec_helper'
describe 'users::test', :type => :class do
  [
    { :operatingsystem => 'Darwin',
      :osfamily => 'Darwin',
      :kernel => 'Darwin',
    },
    { :operatingsystem => 'CentOS',
      :osfamily => 'RedHat',
      :kernel => 'Linux',
    },
    { :operatingsystem => 'Solaris',
      :osfamily => 'Solaris',
      :kernel => 'SunOS',
    },
  ].each do |f|
    context "On #{f[:operatingsystem]}" do
      let(:facts) { f }
      it { should compile } # <-- this fails on the Solaris and Linux cases when run on a Mac
      it { should contain_user('testuser') } # <-- this works all all three platforms
    end
  end  
end
```

When these tests are run on a Mac, I get output like this:

```
$ rspec --color --format d spec/classes/users_test_spec.rb

users::test
  On Darwin
    should compile into a catalogue without dependency cycles
    should contain User[testuser]
  On CentOS
    should compile into a catalogue without dependency cycles (FAILED - 1)
    should contain User[testuser]
  On Solaris
    should compile into a catalogue without dependency cycles (FAILED - 2)
    should contain User[testuser]

Failures:

  1) users::test On CentOS should compile into a catalogue without dependency cycles
     Failure/Error: it { should compile } # <-- this fails on the Solaris and Linux cases when run on a Mac
       error during compilation: Parameter managehome failed on User[testuser]: User provider directoryservice can not manage home directories at modules/users/spec/fixtures/modules/users/manifests/test.pp:11
     # ./spec/classes/users_test_spec.rb:20:in `block (4 levels) in <top (required)>'

  2) users::test On Solaris should compile into a catalogue without dependency cycles
     Failure/Error: it { should compile } # <-- this fails on the Solaris and Linux cases when run on a Mac
       error during compilation: Parameter managehome failed on User[testuser]: User provider directoryservice can not manage home directories at modules/users/spec/fixtures/modules/users/manifests/test.pp:11
     # ./spec/classes/users_test_spec.rb:20:in `block (4 levels) in <top (required)>'

Finished in 0.6555 seconds
6 examples, 2 failures

Failed examples:

rspec ./spec/classes/users_test_spec.rb:20 # users::test On CentOS should compile into a catalogue without dependency cycles
rspec ./spec/classes/users_test_spec.rb:20 # users::test On Solaris should compile into a catalogue without dependency cycles
```

As seen in the Failures section, the provider 'directoryservice' is being used on all three platforms. I'm able to manually specify an appropriate provider for the User type for each platform and this problem goes away.

Finally, here are the Gems I'm using for this test. I tested with rspec-puppet 2.0.0 and 1.0.1

```
$ bundle
Using rake 10.4.2
Using CFPropertyList 2.2.8
Using builder 3.2.2
Using ci_reporter 2.0.0
Using rspec-core 2.99.2
Using diff-lcs 1.2.5
Using rspec-expectations 2.99.2
Using rspec-mocks 2.99.3
Using rspec 2.99.0
Using ci_reporter_rspec 1.0.0
Using facter 2.3.0
Using git 1.2.9.1
Using json_pure 1.8.2
Using hiera 1.3.4
Using metaclass 0.0.4
Using mocha 1.1.0
Using puppet 3.7.3
Using puppet-lint 1.1.0
Using puppet-syntax 1.4.1
Using rspec-puppet 2.0.0
Using puppetlabs_spec_helper 0.8.2
Using bundler 1.7.4
```

It would be nice if the osfamily and operatingsystem facts were respected by the compile matcher at least in terms of default provider selection.

**Comment by mindlace** [comment]
I can confirm this problem with Package[] as well, also on os x.

**Comment by queeno** [comment]
+1

On OS X, the Package[] resource always attempts to use the `pkgdmg` provider, regardless of the overridden default facts, when running `it { should compile }`

I came across this issue when passing a version number to the ensure param in the [Package[sensu] resource](https://github.com/sensu/sensu-puppet/blob/master/manifests/package.pp#L36). In fact, the `pkgdmg` provider is not versionable.

```
error during compilation: Parameter ensure failed on Package[sensu]: Provider must have features 'versionable' to set 'ensure' to '0.20.0-1'
```

I came up with the same workaround described by @geoffdavis, setting a default provider for the Package resource in the pre_condition section:

```
  let(:pre_condition){
    """
    Package {
      provider => 'yum',
    }
    """
  }
```

**Comment by xaniasd** [comment]
This workaround doesn't seem to work with the elasticsearch module (https://github.com/elastic/puppet-elasticsearch), it passes undef as value for the provider attribute in the code..

**Comment by queeno** [comment]
Hey @xaniasd 

Unfortunately adding a default provider in the Package resource won't work, if the elasticsearch module sets it to undef.

I believe you have a couple of options:

- You could attempt to induce rspec-puppet into believing that the running OS is not OSX adding something like this in your `spec_helper.rb` file:
  
  ```
  RSpec.configure do |c|
    c.default_facts = {
      :operatingsystem        => 'CentOS',
      :operatingsystemrelease => '6.6',
      :osfamily               => 'RedHat',
    }
  end
  ```
  This might not work though, because I believe puppet picks the default provider for the package type calling the Facter module directly. In that case, you might consider ways of tricking facter to return a different OS name. However this might break other modules.

- Another option you have is to fork the elasticsearch module and edit it so it never explicitly creates package resources with `provider => undef`.
  
  Instead this:
  ```
    if ($::operatingsystem == 'OpenSuSE') {
      $pkg_provider = 'rpm'
    } else {
      $pkg_provider = undef
    }
  ```
  you could completely scrap the `$pkg_provider` variable and set a provider default, if the operating system is OpenSuSE. The code will also look much neater! (I might open a pull request with them if I have some time later ;) )
  
  ```
  if ($::operatingsystem == 'OpenSuSE') {
    Package {
      provider => 'rpm'
    }
  }
  ```

**Comment by petems** [comment]
I noticed this behaviour with an older version of rspec-puppet, but when I updated to latest head from Github it worked ok? So this may be fixed already

**Comment by petems** [comment]
The diff from the Gemfile.lock:

<img width="319" alt="screenshot 2016-06-09 11 19 51" src="https://cloud.githubusercontent.com/assets/1064715/15926469/21d34788-2e34-11e6-9ecf-b91ebfb96c0a.png">

**Comment by hashar** [comment]
To get the User managedhome statement to pass on Mac OS X, I ended up with:
```
context "On Debian" do
    let(:pre_condition) do
        """
        User {
            provider => 'useradd',
        }
        """  
    end 
    it { should compile }
end     
```

**Comment by DavidS** [comment]
This might also be connected to https://tickets.puppetlabs.com/browse/SDK-20 - I'm pretty sure that providers should not be touched by rspec-puppet.

**Comment by rodjek** [comment]
This should now be resolved by #498
